x/tokenfactory/types/msgs_test.go

```diff
@@ -62,3 +65,408 @@ func TestAuthzMsg(t *testing.T) {
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "invalid subdenom",
+			msg: createMsg(func(msg types.MsgCreateDenom) types.MsgCreateDenom {
+				msg.Subdenom = "thissubdenomismuchtoolongasdkfjaasdfdsafsdlkfnmlksadmflksmdlfmlsakmfdsafasdfasdf"
+				return msg
+			}),
+			expectPass: false,
+		},
+	}
+
+	for _, test := range tests {
+		if test.expectPass {
+			require.NoError(t, test.msg.ValidateBasic(), "test: %v", test.name)
+		} else {
+			require.Error(t, test.msg.ValidateBasic(), "test: %v", test.name)
+		}
+	}
+}
+
+// TestMsgMint tests if valid/invalid create denom messages are properly validated/invalidated
+func TestMsgMint(t *testing.T) {
+	// generate a private/public key pair and get the respective address
+	pk1 := ed25519.GenPrivKey().PubKey()
+	addr1 := sdk.AccAddress(pk1.Address())
+
+	// make a proper mint message
+	createMsg := func(after func(msg types.MsgMint) types.MsgMint) types.MsgMint {
+		properMsg := *types.NewMsgMint(
+			addr1.String(),
+			sdk.NewCoin("bitcoin", sdk.NewInt(500000000)),
+		)
+
+		return after(properMsg)
+	}
+
+	// validate mint message was created as intended
+	msg := createMsg(func(msg types.MsgMint) types.MsgMint {
+		return msg
+	})
+	require.Equal(t, msg.Route(), types.RouterKey)
+	require.Equal(t, msg.Type(), "tf_mint")
+	signers := msg.GetSigners()
+	require.Equal(t, len(signers), 1)
+	require.Equal(t, signers[0].String(), addr1.String())
+
+	tests := []struct {
+		name       string
+		msg        types.MsgMint
+		expectPass bool
+	}{
+		{
+			name: "proper msg",
+			msg: createMsg(func(msg types.MsgMint) types.MsgMint {
+				return msg
+			}),
+			expectPass: true,
+		},
+		{
+			name: "empty sender",
+			msg: createMsg(func(msg types.MsgMint) types.MsgMint {
+				msg.Sender = ""
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "zero amount",
+			msg: createMsg(func(msg types.MsgMint) types.MsgMint {
+				msg.Amount = sdk.NewCoin("bitcoin", sdk.ZeroInt())
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "negative amount",
+			msg: createMsg(func(msg types.MsgMint) types.MsgMint {
+				msg.Amount.Amount = sdk.NewInt(-10000000)
+				return msg
+			}),
+			expectPass: false,
+		},
+	}
+
+	for _, test := range tests {
+		if test.expectPass {
+			require.NoError(t, test.msg.ValidateBasic(), "test: %v", test.name)
+		} else {
+			require.Error(t, test.msg.ValidateBasic(), "test: %v", test.name)
+		}
+	}
+}
+
+// TestMsgBurn tests if valid/invalid create denom messages are properly validated/invalidated
+func TestMsgBurn(t *testing.T) {
+	// generate a private/public key pair and get the respective address
+	pk1 := ed25519.GenPrivKey().PubKey()
+	addr1 := sdk.AccAddress(pk1.Address())
+
+	// make a proper burn message
+	createMsg := func(after func(msg types.MsgBurn) types.MsgBurn) types.MsgBurn {
+		properMsg := *types.NewMsgBurn(
+			addr1.String(),
+			sdk.NewCoin("bitcoin", sdk.NewInt(500000000)),
+		)
+
+		return after(properMsg)
+	}
```

Comment:
Conisder replacing callbacks with just a `baseMsg` to reduce complexity:
```
baseMsg := types.NewMsgBurn(
			addr1.String(),
			sdk.NewCoin("bitcoin", sdk.NewInt(500000000)),
```


so that to customize it we can:
```
{
	name: "empty sender",
	msg: func() {
		msg := *baseMsg
                // change
		return msg
	}),
	expectPass: false,
},
```

Reply:
Great suggestion, changed!